### PR TITLE
endpoint: pod ipvlan support

### PR DESCRIPTION
**Pull request by borkmann**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6608)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

File: api/v1/openapi.yaml [hunk 1]
```diff
@@ -831,6 +831,9 @@ definitions:
       k8s-namespace:
         description: Kubernetes namespace name
         type: string
+      map-id:
```
Comment (tgraf) [inline api/v1/openapi.yaml]: Please call this `policy-map-id`

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -125,7 +126,8 @@ get_ct_map4(struct ipv4_ct_tuple *tuple)
 #if defined ENABLE_IPV4 || defined ENABLE_IPV6
 static inline bool redirect_to_proxy(int verdict, int dir)
 {
-	return verdict > 0 && (dir == CT_NEW || dir == CT_ESTABLISHED);
+	return !is_defined(DATAPATH_IPVLAN) && verdict > 0 &&
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: I suggest we make this more generic and use `!is_defined(ENABLE_PROXY_REDIRECT)` and define that in `common.h` for `DATAPATH_IPVLAN`. Some of the user of `DATAPLANE_IPVLAN` will overlap with what @aanm's flannel PR will need as well. If we use fine-grained defines and select them with the high level options that will make it easier to merge overall.

File: bpf/bpf_lxc.c [hunk 3]
```diff
@@ -176,7 +178,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	 * entry for destination endpoints where we can't encode the state in the
 	 * address.
 	 */
-	if ((svc = lb6_lookup_service(skb, &key)) != NULL) {
+	if (!is_defined(DATAPATH_IPVLAN) && (svc = lb6_lookup_service(skb, &key)) != NULL) {
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: Why are we skipping load balancing?

File: bpf/bpf_lxc.c [hunk 4]
```diff
@@ -328,7 +330,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	}
 
 	/* The packet goes to a peer not managed by this agent instance */
-#ifdef ENCAP_IFINDEX
+#if defined(ENCAP_IFINDEX) && !defined(DATAPATH_IPVLAN)
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: We can use `option.Config.Tunnel = option.TunnelDisabled` in the agent when we run in IPVLAN mode and it will result in `ENCAP_IFINDEX` not being set.

File: bpf/bpf_lxc.c [hunk 5]
```diff
@@ -358,7 +360,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	}
 #endif
 
-#ifdef ENABLE_NAT46
+#if defined(ENABLE_NAT46) && !defined(DATAPATH_IPVLAN)
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: Let's do `#undefine ENABLE_NAT46` in `common.h` when `DATAPATH_IPVLAN` is set

File: bpf/bpf_lxc.c [hunk 6]
```diff
@@ -367,7 +369,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	goto pass_to_stack;
 
 to_host:
-	if (1) {
+	if (!is_defined(DATAPATH_IPVLAN)) {
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: Let's add `USE_HOST_REDIRECT`

File: bpf/bpf_lxc.c [hunk 7]
```diff
@@ -730,19 +735,18 @@ int handle_ingress(struct __sk_buff *skb)
...
 		break;
 #endif
-
 #ifdef ENABLE_IPV4
 	case bpf_htons(ETH_P_IP):
 		ep_tail_call(skb, CILIUM_CALL_IPV4_FROM_LXC);
 		ret = DROP_MISSED_TAIL_CALL;
 		break;
-
+#if !defined(DATAPATH_IPVLAN)
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `ENABLE_ARP_RESPONDER`

File: bpf/bpf_lxc.c [hunk 8]
```diff
@@ -1089,7 +1093,7 @@ __section_tail(CILIUM_MAP_POLICY, LXC_ID) int handle_policy(struct __sk_buff *sk
 	return ret;
 }
 
-#ifdef ENABLE_NAT46
+#if defined(ENABLE_NAT46) && !defined(DATAPATH_IPVLAN)
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: This can stay as-is if we undefine `ENABLE_NAT46`

**Review by tgraf — CHANGES_REQUESTED** [review]
First review of the datapath portion as that will conflict the most with the flannel work

File: bpf/bpf_lxc.c (hunk 2 above)
Comment (borkmann) [reply bpf/bpf_lxc.c]: Yes, fully agree, will fix is up this way.

File: bpf/bpf_lxc.c (hunk 4 above)
Comment (borkmann) [reply bpf/bpf_lxc.c]: Agree, I was planning to remove this here anyway like I did in bpf_netdev because tunnel mode needs to be disabled in order to select ipvlan data path mode, so this is not needed.

File: api/v1/openapi.yaml (hunk 1 above)
Comment (borkmann) [reply api/v1/openapi.yaml]: Changed it into data-path-map-id since this denotes the tail call map for the pod that cilium is managing where it attaches bpf_lxc onto, so only policy-map-id would be too confusing as well.

File: bpf/bpf_lxc.c (hunk 8 above)
Comment (borkmann) [reply bpf/bpf_lxc.c]: Agree, I'll improve this logic. NAT64 code doesn't work with ipvlan since if we rewrite addresses from v4 to v6 and vice versa, then ipvlan internal fdb logic won't find local target container anymore. If we ever need this feature here we could i) push the skbs to the phys device, and do reverse translation from there to then push it back in so that lookup would work, or we could add a ipvlan bpf operation mode to have custom lookups (or even policy moved into it).

Comment (borkmann) [reply bpf/bpf_lxc.c]: I've changed it to error out if a user tries to flip the ENABLE_NAT46 via `cilium endpoint config`.

File: bpf/bpf_lxc.c (hunk 3 above)
Comment (borkmann) [reply bpf/bpf_lxc.c]: Ok, I've removed it. I had this originally to avoid another lookup in fast path but also because LB for local services does not work yet, but we can document this under current limitations for initial merge.

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-me-please

File: Vagrantfile [hunk 9]
```diff
@@ -119,7 +119,7 @@ Vagrant.configure(2) do |config|
         vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
         vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
 
-        config.vm.box = "cilium/ubuntu-dev"
+        config.vm.box = "cilium/ubuntu"
```
Comment (tgraf) [inline Vagrantfile]: This change seems unrelated

File: api/v1/openapi.yaml [hunk 10]
```diff
@@ -1407,9 +1410,20 @@ definitions:
       deviceMTU:
         description: MTU on workload facing devices
         type: integer
+      deviceIfIndex:
```
Comment (tgraf) [inline api/v1/openapi.yaml]: This is a bit too generic. We want to get rid of the `-d` argument if possible as we have no way of passing it in generically for all worker nodes in a cluster. Let's call this `ipvlan-master-device-index` or similar to make it clear what it is.

File: pkg/endpoint/endpoint.go [hunk 11]
```diff
@@ -136,6 +141,9 @@ type Endpoint struct {
 	// libnetwork
 	DockerEndpointID string
 
+	// Corresponding BPF map identifier
+	mapID int
```
Comment (tgraf) [inline pkg/endpoint/endpoint.go]: Please use a more specific name

File: pkg/option/config.go [hunk 12]
```diff
@@ -441,6 +444,7 @@ type DaemonConfig struct {
 	RunDir          string     // Cilium runtime directory
 	NAT46Prefix     *net.IPNet // NAT46 IPv6 Prefix
 	Device          string     // Receive device
+	DeviceIfIndex   int        // Device interface index
```
Comment (tgraf) [inline pkg/option/config.go]: Let's make this IPVLAN specific

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks great. A couple of API facing comments that we should address before merging as it becomes harder to change. The rest seems good to go to get initial user testing coverage.

**Comment by brb** [comment]
test-me-please

File: api/v1/openapi.yaml (hunk 10 above)
Comment (borkmann) [reply api/v1/openapi.yaml]: The above `deviceIfIndex` is basically only internal information, meaning the CNI queries the Cilium daemon for the specified ipvlan master device in order to setup ipvlan slaves bound to it. We could rename the above, or are you saying that the above property name is okay, just that the daemon command line parameter should be renamed to something as specific as `--ipvlan-master-device-index`? I actually liked reusing the `--device` param, so that it generically becomes `--datapath-mode ipvlan --device eth0`. Even if we get rid of `--device`, wouldn't we still have the same issue for ipvlan at that point? Meaning, if we have some generic master device detection in place, wouldn't this be needed for both datapath flavors?

**Comment by brb** [comment]
test-me-please

File: api/v1/openapi.yaml (hunk 10 above)
Comment (tgraf) [reply api/v1/openapi.yaml]: The above change to `openapi.yaml` becomes stable API. We can never change it again. If it refers to the IPVLAN master ifindex then let's label it as such.

The main difference between prior use of `--device` that it referred to an ethernet device name which can be different on each worker node depending on HW or node configuration whereas the IPVLAN master device is likely more stable with something such as `ipvlan0`.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

File: api/v1/openapi.yaml (hunk 10 above)
Comment (borkmann) [reply api/v1/openapi.yaml]: Ok, I've changed it and added a comment for rationale to keep it separate.

File: pkg/option/config.go [hunk 13]
```diff
@@ -245,6 +245,12 @@ const (
 	// MTUName is the name of the MTU option
 	MTUName = "mtu"
 
+	// DatapathMode is the name of the DatapathMode option
+	DatapathMode = "datapath-mode"
+
+	// DatapathMode is the name of the DatapathMode option
```
Comment (tgraf) [inline pkg/option/config.go]: Small comment change left but we can do that after the CI passes if this is the only change left.

**Review by tgraf — APPROVED** [review]
Minor comment change left that we can do after the CI passes and then merge

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
Test timed out, retrying.

**Comment by borkmann** [comment]
test-me-please
